File: internal/pkg/ratelimitx/leaky_bucket.go
package ratelimitx

import (
	"math"
	"time"
)

type memLeakyBucket struct {
	level float64
	last  time.Time
}

func (m *memLimiter) allowLeakyBucket(key string, limit int64, window time.Duration, now time.Time) (bool, int64) {
	capacity := float64(limit)
	item, ok := m.leaks[key]
	if !ok {
		item = memLeakyBucket{last: now}
	} else {
		elapsed := now.Sub(item.last)
		if elapsed < 0 {
			elapsed = 0
		}
		item.level = math.Max(0, item.level-elapsed.Seconds()*capacity/window.Seconds())
		item.last = now
	}

	if item.level+1 > capacity {
		m.leaks[key] = item
		return false, floorRemaining(capacity - item.level)
	}
	item.level++
	m.leaks[key] = item
	return true, floorRemaining(capacity - item.level)
}

File: internal/pkg/ratelimitx/scripts/leaky_bucket.lua
local key = KEYS[1]
local now = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local limit = tonumber(ARGV[3])
local capacity = limit

local level = tonumber(redis.call("HGET", key, "level"))
local ts = tonumber(redis.call("HGET", key, "ts"))

if level == nil then
    level = 0
end
if ts == nil then
    ts = now
end

local elapsed = now - ts
if elapsed < 0 then
    elapsed = 0
end

level = math.max(0, level - elapsed * capacity / window)

local allowed = 0
if level + 1 <= capacity then
    level = level + 1
    allowed = 1
end

local remaining = math.floor(capacity - level)
if remaining < 0 then
    remaining = 0
end

redis.call("HSET", key, "level", level, "ts", now)
redis.call("PEXPIRE", key, window * 2)

return {allowed, remaining}
